### DISCUSSION TITLE
Type of parts of the AST in Typescript unclear

### DISCUSSION THREAD

**Discussion by lkj4** [syntax-tree/unist#51] [Q&A] [answered]
After getting a chunk (an array of nodes, to be clear) of nodes with `unist-util-find-all-between` in an AST I try to convert that chunk back to Markdown with `mdast-util-to-markdown` but IDK how. I tried embedding that chunk  into a...

```
{ 
  type: "root",
  children: chunk 
}
```
and got 
```
[tsserver 2345] [E] Argument of type 'object' is not assignable to parameter of type 'Node'. 
Type '{}' is missing the following properties from type 'FootnoteReference': type, identifier
```
then I asserted that chunk `as Node` and got
```
[tsserver 2345] [E] Argument of type 'Node<Data>' is not assignable to parameter of type 'Node'. 
Property 'identifier' is missing in type 'Node<Data>' but required in type 'FootnoteReference'.
```
then I asserted it `as Node<Data>` with `Data` being...
```
interface Data {
  identifier: string
}
```
and got the same again...
```
[tsserver 2345] [E] Argument of type 'Node<Data>' is not assignable to parameter of type 'Node'.
Property 'identifier' is missing in type 'Node<Data>' but required in type 'FootnoteReference'.
```

So, how I can I deal with AST chunks with TS?

**Comment by lkj4** [comment]
Think I found the correct type for that chunk embedded into a Root type: `Root` from `mdast` haha. Nice and wondering if I need to embed that chunk at all...

**Reply by ChristianMurphy** [reply]
probably not,
that cast sounds off, `chunk` is `Node[]` right? A list of `Node`. `Root` is just a single object.
Forcing TS to type cast a list of object to an object may make the compile time error go away, but may cause a runtime error.

**Reply by lkj4** [reply]
Oh yes, right `chunk` is `Node[]`. Just tried that too without embedding the nodes into that fake root and also `Node<Data>[]` paired with the `interface Data` from above and got...
```
[tsserver 2345] [E] Argument of type 'Node<Data>[]' is not assignable to
parameter of type 'Node'.
  Type 'Node<Data>[]' is missing the following properties from type
'FootnoteReference': type, identifier
```

**Reply by ChristianMurphy** [reply]
Right, most plugins are designed to take a tree, starting with a single node.
Not a list of nodes.

TS is warning about this, in a somewhat round about way, it knows `Array` does not have the `type` or `identifier` properties which must exist on a `Node`.

**Comment by lkj4** [comment]
Hm, `mdast` is deprecated and I should use `remark` but latter does not export a type `Root`...

**Reply by ChristianMurphy** [reply]
`mdast` was a JavaScript package which has since been deprecated and replaced by remark.
`@types/mdast` is a TypeScript typings package which is used by much of the remark ecosystem.

**Comment by lkj4** [comment]
Ok, I saw in some of the docs that I should import `@types/mdast` but I actually don't need to. Even without importing the types, I am able to do an `import type { Root } from 'mdast'` which is a bit confusing. So what is actually happening here, where are the types actually coming from? I do not have `mdast` installed and use `pnpm` which does not give access to indirect modules.

Btw, and re `List`, TS does not complain about `List` as assertion but it creates an RT error, so `toMarkdown` requires `Root`.

**Reply by ChristianMurphy** [reply]
> Even without importing the types, I am able to do an `import type { Root } from 'mdast'` which is a bit confusing. So what is actually happening here, where are the types actually coming from?

Probably a transitive dependency.
Many `remark=*` packages include `@types/mdast` as a dependency. For example
https://github.com/remarkjs/remark/blob/383871e6307fa0c09b542a78c05974267b759fe4/packages/remark-parse/package.json#L44
yarn and recent versions of `npm` flatten dependencies.
meaning `./node_modules/@types/mdast` will be present even if you haven't directly run `npm install @types/mdast`  or `yarn add @types/mdast`

**Reply by lkj4** [reply]
Yeah right, however [`pnpm` does not flatten the packages](https://pnpm.io/motivation#creating-a-non-flat-node_modules-directory), so it shouldn't allow access. I could imagine though that it might be different with types...

**Comment by ChristianMurphy** [comment, marked as answer]
@lkj4 this incompatibility is likely coming from

https://github.com/mrzmmr/unist-util-find-all-between/blob/adb9565a4424bcd9bcd3db4199b4d15ba91c6ecb/types/index.d.ts#L16-L21

`remark-*` plugins expect to get mdast nodes, this is returning a list more generic unist `Node`, which doesn't have the properties generally expected on mdast.

---

a few options:
1, open an issue (or even better a PR) with https://github.com/mrzmmr/unist-util-find-all-between to add generics and type narrowing so it can figure out the type it is passed.
2. share a sample of the code you are trying, so folks can offer more concrete suggestions on what to do

In the meantime, you may want to checkout https://unifiedjs.com/learn/guide/syntax-trees-typescript/ and https://unifiedjs.com/learn/recipe/narrow-node-typescript/ which introduce the unified syntax trees and some of the approaches to getting the specific types.

**Reply by lkj4** [reply]
Thanks for sharing the links, let get me a bit more familiar with ASTs and then I might submit issue/PR, right now the fake root hack works ok.
